code/modules/particles/byond_particles/particool.dm

```diff
@@ -0,0 +1,253 @@
+	box			| vector | A random vector within a box whose corners are at A and B.
+	color		  | color  | (string) or color matrix	Result type depends on whether A or B are matrices or not. The result is interpolated between A and B; components are not randomized separately.
+	circle		 | vector | A random XY-only vector in a ring between radius A and B, centered at 0,0.
+	sphere		 | vector | A random vector in a spherical shell between radius A and B, centered at 0,0,0.
+	square		 | vector | A random XY-only vector between squares of sizes A and B. (The length of the square is between A*2 and B*2, centered at 0,0.)
+	cube		   | vector | A random vector between cubes of sizes A and B. (The length of the cube is between A*2 and B*2, centered at 0,0,0.)
+	*/
+
+	// I write code like this because I hate myself
+	var/a = length(stringToList(L["a"],TRUE)) > 1 ? stringToList(L["a"],TRUE) : text2num(L["a"])
+	var/b = length(stringToList(L["b"],TRUE)) > 1 ? stringToList(L["b"],TRUE) : text2num(L["b"])
+
+
+	var/rand_type = parse_rand_type(L["rand"])
+
+	switch(L["genType"])
+		if("num")	return generator(L["genType"], a, b, rand_type)
+		if("vector") return generator(L["genType"], a, b, rand_type)
+		if("box")	return generator(L["genType"], a, b, rand_type)
+		if("color") //Color can be string or matrix
+			a = length(a) > 1 ? a : L["a"]
+			b = length(b) > 1 ? b : L["b"]
+			return generator(L["genType"], a, b, rand_type)
+		if("circle") return generator(L["genType"], a, b, rand_type)
+		if("sphere") return generator(L["genType"], a, b, rand_type)
+		if("square") return generator(L["genType"], a, b, rand_type)
+		if("cube")   return generator(L["genType"], a, b, rand_type)
+	return null
+
+/datum/particle_editor/proc/parse_rand_type(rand_type)
+	switch(rand_type)
+		TEXT_TO_MACRO(UNIFORM_RAND)
+		TEXT_TO_MACRO(LINEAR_RAND)
+		TEXT_TO_MACRO(NORMAL_RAND)
+		TEXT_TO_MACRO(SQUARE_RAND)
+	CRASH("Unknown rand type [rand_type]")
+
+/datum/particle_editor/proc/debugOutput(L, nodeName)
+	if(istype(L,/list))
+		for(var/elem in L)
+			if(istype(elem,/list))
+				log_game(nodeName)
+				debugOutput(elem, nodeName + ":LIST" + elem)
+			else
+				log_game(nodeName + ":" + elem + ":" + text("[]",elem))
+	else
+		log_game(L)
+
+/datum/particle_editor/ui_act(action, list/params)
+	. = ..()
+	if(.)
+		return
+
+	debugOutput(params)
+
+	switch(action)
+		if("add_particle")
+			target.add_particle()
+			. = TRUE
+		if("remove_particle")
+			target.remove_particle()
+			. = TRUE
+		if("modify_particle_value")
+			target.modify_particle_value(params["new_data"]["name"], translate_value(params["new_data"]))
+			. = TRUE
+		if("modify_color_value")
+			var/new_color = input(usr, "Pick new particle color", "Particool Colors!") as color|null
+			if(new_color)
+				target.modify_particle_value("color",new_color)
+				. = TRUE
+		if("modify_icon_value")
+			var/icon/new_icon = input("Pick icon:", "Icon") as null|icon
+			if(new_icon && target.particles)
```

Comment:
ditto

Reply:
input() as null | icon should sanitize it enough